lib/client/boluscalc.js

```diff
@@ -0,0 +1,776 @@
+      merged.add(1, 'hours');
+    }
+    if (ele.attr('oldminutes') === '0' && merged.minutes() === 59) {
+      merged.add(-1, 'hours');
+    }
+    
+    setDateAndTime(merged);
+    updateTime(ele, merged);
+    boluscalc.eventTimeTypeChange();
+    
+    // update BG from sgv to this time
+    setBG(findClosestSGVToPastTime(merged.toDate()), merged.toDate());
+
+    boluscalc.calculateInsulin();
+    maybePrevent(event);
+//    Nightscout.utils.updateBrushToTime(moment.toDate());
+  };
+
+  boluscalc.eventTimeTypeChange = function eventTimeTypeChange(event) {
+    if ($('#bc_othertime').is(':checked')) {
+      $('#bc_eventTimeValue').focus();
+      $('#bc_retro').css('display','');
+      if (mergeDateAndTime()<moment()) {
+        $('#bc_retro').css('background-color','red').text(translate('RETRO MODE'));
+      } else if (mergeDateAndTime()>moment()) {
+        $('#bc_retro').css('background-color','blue').text(translate('IN THE FUTURE'));
+      } else {
+        $('#bc_retro').css('display','none');
+      }
+    } else {
+      $('#bc_retro').css('display','none');
+      setDateAndTime();
+      boluscalc.updateVisualisations(boluscalc.lastsbx);
+      boluscalc.calculateInsulin();
+//        Nightscout.utils.setYAxisOffset(50); //50% of extend
+//        Nightscout.utils.updateBrushToTime(Nightscout.utils.mergeInputTime($('#bc_eventTimeValue').val(), $('#bc_eventDateValue').val()).toDate());
+    }
+    maybePrevent(event);
+  };
+
+  boluscalc.toggleDrawer = function toggleDrawer (event) {
+    boluscalc.prepare();
+    client.browserUtils.toggleDrawer('#boluscalcDrawer');
+    maybePrevent(event);
+  };
+
+  boluscalc.prepare = function prepare( )  {
+    foods = [];
+    $('#bc_profile').empty();
+    client.profilefunctions.listBasalProfiles().forEach(function (p) {
+      $('#bc_profile').append('<option val="' + p + '">' + p + '</option>');
+    });
+    $('#bc_usebg').prop('checked','checked');
+    $('#bc_usecarbs').prop('checked','checked');
+    $('#bc_usecob').prop('checked','');
+    $('#bc_useiob').prop('checked','checked');
+    $('#bc_bgfromsensor').prop('checked','checked');
+    $('#bc_carbs').val('');
+    $('#bc_quickpick').val(-1);
+    $('#bc_preBolus').val(0);
+    $('#bc_notes').val('');
+    $('#bc_enteredBy').val($.localStorage.get('enteredBy') || '');
+    $('#bc_nowtime').prop('checked', true);
+    $('#bc_othercorrection').val(0);
+    $('#bc_profile').val(client.profilefunctions.activeProfileToTime());
+    setDateAndTime();
+    boluscalc.eventTimeTypeChange();
+    if (boluscalc.lastsbx) {
+      boluscalc.updateVisualisations(boluscalc.lastsbx);
+    }
+    boluscalc.calculateInsulin();
+  };
+  
+  boluscalc.calculateInsulin = function calculateInsulin (event) {
+    maybePrevent(event);
+    boluscalc.gatherBoluscalcData( );
+    boluscalc.updateGui(boluscalc.record);
+    return  boluscalc.record;
+  };
+  
+  boluscalc.updateGui = function updateGui (record) {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1270838365/issues/source?bid=2452379&fileBranchId=2935247#l184)